clients/common/host_alloc.cpp

```diff
@@ -0,0 +1,229 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ * ************************************************************************ */
+
+#ifdef WIN32
+#include <windows.h>
+
+#else
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+#endif
+
+#include <map>
+#include <mutex>
+#include <stdlib.h>
+
+#include "hipblas_test.hpp"
+#include "host_alloc.hpp"
+
+// light weight memory tracking for threshold limit on total use
+static size_t                  mem_used{0};
+static std::map<void*, size_t> mem_allocated;
+static std::mutex              mem_mutex;
+
+inline void alloc_ptr_use(void* ptr, size_t size)
+{
+    std::lock_guard<std::mutex> lock(mem_mutex);
+    if(ptr)
+    {
+        mem_allocated[ptr] = size;
+        mem_used += size;
+    }
+}
+
+inline void free_ptr_use(void* ptr)
+{
+    std::lock_guard<std::mutex> lock(mem_mutex);
+    if(ptr && mem_allocated[ptr])
+    {
+        mem_used -= mem_allocated[ptr];
+        mem_allocated.erase(ptr);
+    }
+}
+
+size_t host_bytes_allocated()
+{
+    std::lock_guard<std::mutex> lock(mem_mutex);
+    return mem_used;
+}
+
+//!
+//! @brief Memory free helper.  Returns kB or -1 if unknown.
+//!
+ptrdiff_t host_bytes_available()
+{
+#ifdef WIN32
+
+    MEMORYSTATUSEX status;
+    status.dwLength = sizeof(status);
+    GlobalMemoryStatusEx(&status);
+    return (ptrdiff_t)status.ullAvailPhys;
+
+#else
+
+    const int BUF_MAX = 1024;
+    char      buf[BUF_MAX];
+
+    ptrdiff_t n_bytes = -1; // unknown
+
+    FILE* fp = popen("cat /proc/meminfo", "r");
+    if(fp == NULL)
+    {
+        return n_bytes;
+    }
+
+    static const char* mem_token     = "MemFree";
+    static auto*       mem_free_type = getenv("HIPBLAS_CLIENT_ALLOC_AVAILABLE");
+    if(mem_free_type)
+    {
+        mem_token = "MemAvail"; // MemAvailable
+    }
+    int mem_token_len = strlen(mem_token);
+
+    while(fgets(buf, BUF_MAX, fp) != NULL)
+    {
+        // set env HIPBLAS_CLIENT_ALLOC_AVAILABLE to use MemAvailable if too many SKIPS occur
+        if(!strncmp(buf, mem_token, mem_token_len))
+        {
+            sscanf(buf, "%*s %td", &n_bytes); // kB assumed as 3rd column and ignored
+            n_bytes *= 1024;
+            break;
+        }
+    }
+
+    int status = pclose(fp);
+    if(status == -1)
+    {
+        return -1;
+    }
+    else
+    {
+        return n_bytes;
+    }
+
+#endif
+}
+
+inline bool host_mem_safe(size_t n_bytes)
+{
+#if defined(HIPBLAS_BENCH)
+    return true; // roll out to hipblas-bench when CI does perf testing
+#else
+    static auto* no_alloc_check = getenv("HIPBLAS_CLIENT_NO_ALLOC_CHECK");
+    if(no_alloc_check)
+    {
+        return true;
+    }
+
+    constexpr size_t threshold = 100 * 1024 * 1024; // 100 MB
+
+    static size_t client_ram_limit = 0;
+
+    static int once = [&] {
+        auto* alloc_limit = getenv("HIPBLAS_CLIENT_RAM_GB_LIMIT");
```

Comment:
Rather than this library specific perhaps better to ask CI to set ROCM_CI_CLIENT_RAM_GB_LIMIT as they just added ROCM_CI_RAM_GB_LIMIT for building.  Get away from a variable per math library....

Reply:
I checked it locally and it seems to be working fine. Sure, when we remove the ROCBLAS_CLIENT_RAM_GB_LIMIT will remove the HIPBLAS_CLIENT_RAM_GB_LIMIT as well, or do you want me to remove this feature from this PR?

Reply:
Fine to leave in for now and change later with rocblas, does the groovy set this var?

Reply:
No the groovy doesn't set it.

Reply:
Well set the env in the groovy so your code is tested by CI.  Can be this PR or another as simple additions